File: lib/models/src/PropertyValueDefinition.ts
import {NamedModel} from "./NamedModel";
import {RDF, RDFS} from "./vocabularies";
import {Term} from "n3";
import {Image} from "./Image";
import {ThumbnailSelector} from "./ThumbnailSelector";
import {selectThumbnail} from "./selectThumbnail";

export class PropertyValueDefinition extends NamedModel {
  get images(): readonly Image[] {
    return this.dataset.imagesByDepictsUri(this.uri);
  }

  get label(): string | null {
    return this.optionalString(RDFS.label);
  }

  get propertyUris(): readonly string[] {
    const propertyNamedNodes = this.parentNamedNodes(RDF.predicate);
    if (propertyNamedNodes.length === 0) {
      throw new RangeError(
        "property value definition must link to one or more property definitions"
      );
    }
    return propertyNamedNodes.map(node => node.value);
  }

  thumbnail(selector: ThumbnailSelector): Image | null {
    return selectThumbnail(this.images, selector);
  }

  get value(): Term {
    return this.requiredTerm(RDF.value);
  }
}

File: paradicms_etl/models/property_value_definition.py
from typing import Optional, Tuple

from rdflib.namespace import RDF, RDFS
from rdflib.term import Literal, Node, URIRef

from paradicms_etl.models._named_model import _NamedModel


class PropertyValueDefinition(_NamedModel):
    @classmethod
    def from_fields(
        cls,
        *,
        property_uris: Tuple[URIRef, ...],  # Child -> parent references
        uri: URIRef,
        value: Node,  # A property value
        label: Optional[str] = None,
    ):
        resource = cls._create_resource(identifier=uri)
        if label is not None:
            resource.add(RDFS.label, Literal(label))
        if not property_uris:
            raise ValueError("must specify at least one property URI")
        for property_uri in property_uris:
            # Not kosher use of rdf:predicate, but I couldn't find a better fit
            resource.add(RDF.predicate, property_uri)
        resource.add(RDF.value, value)
        return cls(resource)
